Add `sample_rate` and `dmin` arguments to `DataSet.to_reciprocalgrid()`

JBGreisman opened the pull request:
This PR addresses #175 by updating the `DataSet.to_reciprocalgrid()` call signature to accept `sample_rate` and `dmin` arguments. The default value for `sample_rate=3.0` and the default for `dmin=None`, which corresponds to using the full resolution of the `DataSet` instance. The `to_reciprocalgrid()` method can still accept an explicit `gridsize`,  and if one is provided, it supersedes the `sample_rate` and `dmin` arguments. 

To facilitate this method, I also added `rs.utils.get_gridsize()`, which can be used for calculating a proper gridsize for a `DataSet` instance. 

Lastly, this PR cleans up and adds additional tests to `test_dataset_grid.py` to more thoroughly test this function.

kmdalton commented:
I think a `get_fourier_gridsize` type of function should exist in `utils` because it seems generally useful especially in the context of initializing deep learning models. I don't think _this_ `get_gridsize` function belongs in `utils` though. In my mind nothing in `utils` should require a `dataset` instance. Let's do this at the top level for now. If you want a separate method to handle the grid sizing, I'd suggest writing a "private" `DataSet._get_gridsize` helper. 

I can do a more thorough review tomorrow.

JBGreisman commented:
Thanks for the comment -- I agree that as written the function wasn't right for `rs.utils`, which we try to keep agnostic to any implementation/interface details in DataSet. 

I just pushed an update to this PR that changes the `rs.utils.get_gridsize()`call signature to be more general. I also added a specific helper function `DataSet.get_gridsize()`, which is then called in `DataSet.to_reciprocalgrid()`.

kmdalton commented on reciprocalspaceship/dataset.py:
```diff
@@ -1287,31 +1287,83 @@ def canonicalize_phases(self, inplace=False):
         -----
         - The data being arranged on a reciprocal grid must be compatible
           with a numpy datatype.
+        - Any missing Miller indices are initialized to zero.
```
This is out of scope for this PR, but we should consider initializing by interpolating adjacent grid points. Along these lines though, should there be a `fill_value=0.` option?

JBGreisman replied:
I agree this could be a future change. Given how much is going on in this PR, I'd rather defer any additional changes

kmdalton replied:
Fine with me.

kmdalton commented on reciprocalspaceship/dataset.py:
```diff
@@ -1287,31 +1287,83 @@ def canonicalize_phases(self, inplace=False):
+        - The grid size determined using `sample_rate` and `dmin` will depend
+          on the cell parameters of the dataset. If the grid size must be
+          consistent across different isomorphous cell parameters, `gridsize`
+          can be explicitly provided.
```
explicitly provided. `gridsize` supersedes `dmin` and `sample_rate` for grid size determination.

JBGreisman replied:
updated docstring

kmdalton commented on reciprocalspaceship/dataset.py:
```diff
@@ -1287,31 +1287,83 @@ def canonicalize_phases(self, inplace=False):
+        dmin : float
+            Highest-resolution reflection to consider for grid size. If None,
+            `dmin` will be set to the highest resolution reflection in the
+            dataset (default: None)
```
Need to say: "The reflections used to populate the grid will also be truncated to resolution `>=dmin`. "

JBGreisman replied:
updated docstring

kmdalton commented on reciprocalspaceship/dataset.py:
```diff
@@ -1287,31 +1287,83 @@ def canonicalize_phases(self, inplace=False):
         return self
 
-    def to_reciprocalgrid(self, key, gridsize):
+    def get_gridsize(self, dmin=None, sample_rate=3.0):
```
I propose to change the name of this method to `get_fourier_gridsize` or `get_reciprocal_gridsize` something similar. Otherwise, it is not apparent what the "grid" is. I think the is the same reason it's called `ds.to_reciprocalgrid` and not simply `ds.to_grid`.

kmdalton replied:
This is a nitpick but i think `gridsize` should be `grid_size` in parameter and method names. I don't see why we would contract it.

JBGreisman replied:
updated call signature and method name

kmdalton commented on reciprocalspaceship/utils/grid.py:
```diff
@@ -0,0 +1,55 @@
+
+@cellify
+@spacegroupify
+def get_gridsize(cell, dmin, sample_rate=3.0, spacegroup=None):
```
I'd prefer `get_fourier_grid_size` or `get_reciprocal_grid_size`

JBGreisman replied:
done -- for consistency elsewhere, I've made it `get_reciprocal_grid_size`

kmdalton commented on reciprocalspaceship/utils/grid.py:
```diff
@@ -0,0 +1,55 @@
+
+    Raises
+    ------
+    ValueError
```
Nice. I always forget to do this.

kmdalton commented on tests/test_dataset_grid.py:
```diff
@@ -5,49 +5,71 @@
-def test_to_reciprocalgrid_float(mtz_by_spacegroup, sample_rate, p1):
+@pytest.mark.parametrize("sample_rate", [2.0, 3.0])
+@pytest.mark.parametrize("column", ["FMODEL", "sf"])
+def test_to_reciprocalgrid_p1(mtz_by_spacegroup, sample_rate, column):
```
Not sure this test is necessary, but I guess it doesn't hurt to have it.

JBGreisman replied:
This test is very similar to the ones done for `expand_to_p1()`. Though somewhat repetitive, it seemed like a good idea to ensure that there is no dependency on `sample_rate`

kmdalton commented on tests/test_dataset_grid.py:
```diff
@@ -5,49 +5,71 @@
+@pytest.mark.parametrize("sample_rate", [1.0, 2.5, 3.0])
+@pytest.mark.parametrize("dmin", [10.0, 5.0, None])
+@pytest.mark.parametrize("gridsize", [None, [144, 144, 144]])
+def test_to_reciprocalgrid_sizes(mtz_by_spacegroup, sample_rate, dmin, gridsize):
```
Looks good.

kmdalton commented on tests/utils/test_grid.py:
```diff
@@ -0,0 +1,42 @@
+
+        # shape of output
+        assert isinstance(result, list)
+        assert len(result) == 3
```
as long as we're doing this, you may as well make sure the list contains ints. something like: 
```python
assert all([isinstance(i, int) for i in result])
```

JBGreisman replied:
good idea -- just added this.

kmdalton reviewed (changes requested):
I have some minor suggestions and some semantic concerns. Mostly I think anywhere that says `gridsize` should be `grid_size`, and we should make it clear what grid we're talking about in the method names. Either `get_grid_size` --> `get_fourier_grid_size` or `get_reciprocal_grid_size`. 

Otherwise, this seems like a great PR. As far as I can tell, the tests are well engineered.

JBGreisman commented:
I made the requested call signature and method name changes -- I agree that they all clarify things, and make things more consistent with our naming scheme. 

The big additional API change is that I changed `DataSet.to_reciprocalgrid()`--> `DataSet.to_reciprocal_grid()` to make things as consistent as possible. Since this change will break backwards compatibility, I have kept `DataSet.to_reciprocalgrid(key, gridsize=...)` and added a `DeprecationWarning`. I have also updated the documentation with a warning, and have changed the example notebooks to use the new function. In a future release we can fully remove support for the deprecated version.

kmdalton commented on reciprocalspaceship/dataset.py:
```diff
@@ -3,7 +3,7 @@
 from pandas.api.types import is_complex_dtype
 
-from reciprocalspaceship import concat, dtypes
+from reciprocalspaceship import concat, dtypes, utils
```
Why don't you just import `get_reciprocal_grid_size` in the `utils` function imports below? Wouldn't that be more consistent?

kmdalton reviewed (changes requested):
I think we're nearing completion here. See my comment about imports. 

In example 5 from the docs there is still a reference to `to_reciprocalgrid` in one of the markdown cells. You also use `reciprocalgrid` as a variable name in that example. You might want to update that, but that is pretty much just aesthetic. 

I couldn't find any other issues with misusing `grid_size` or `reciprocalgrid`.

JBGreisman commented:
Made the two changes -- thanks for the review! Please feel free to merge if you think this is good to go

kmdalton commented:
Sorry -- I forgot to point out that the call to `get_reciprocal_grid_size` still had a `utils` prefix. I just pushed a commit to fix. The CI should pass now :crossed_fingers:

JBGreisman commented:
Thanks -- good catch.
